xwiki-pro-macros-api/src/main/java/com/xwiki/macros/tab/internal/TabGroupMacro.java

```diff
@@ -0,0 +1,203 @@
+
+    private static final String CSS_PX_UNIT = "px;";
+
+    private static final String BLOCK_PARAM_CLASS = "class";
+
+    private static final String BLOCK_PARAM_ROLE = "role";
+
+    private static final Random RANDOM = new Random();
+
+    @Inject
+    private Provider<XWikiContext> xwikiContextProvider;
+
+    @Inject
+    private MacroContentParser contentParser;
+
+    @Inject
+    @Named("ssrx")
+    private SkinExtension ssfx;
+
+    /**
+     * Create and initialize the descriptor of the macro.
+     */
+    public TabGroupMacro()
+    {
+        super(NAME, DESCRIPTION, new DefaultContentDescriptor(CONTENT_DESCRIPTION, true, Block.LIST_BLOCK_TYPE),
+            TabGroupMacroParameters.class);
+    }
+
+    @Override
+    public boolean supportsInlineMode()
+    {
+        return false;
+    }
+
+    private List<Block> renderView(TabGroupMacroParameters parameters, String content,
+        MacroTransformationContext context) throws MacroExecutionException
+    {
+        ssfx.use("css/tabmacro.css");
+        String macroId = parameters.getId();
+
+        if (StringUtils.isEmpty(macroId)) {
+            // FIXME we should have a better way to generate an ID
+            macroId = "tab-group-" + RANDOM.nextInt();
+        }
+
+        Block contentBlocks = this.contentParser.parse(content, context, false, context.isInline());
+
+        List<Block> ulBlocks = new ArrayList<>();
+        List<Block> macroBlocks = contentBlocks.getBlocks(new MacroBlockMatcher(TAB_MACRO_ID), Block.Axes.CHILD);
+        Optional<Block> defaultTabBlock = macroBlocks.stream()
+            .filter(i -> TRUE.equals(i.getParameter(TabMacroParameters.PARAM_NAME_DEFAULT)))
+            .findFirst();
+        int defaultTabIndex = defaultTabBlock.isPresent() ? macroBlocks.indexOf(defaultTabBlock.get()) : 0;
+
+        int inc = 0;
+        for (Block mb : macroBlocks) {
+            boolean isActive = inc == defaultTabIndex;
+            mb.setParameter(TabMacroParameters.PARAM_NAME_DEFAULT, isActive ? TRUE : FALSE);
+            String id = mb.getParameter(ID);
+            if (id == null) {
+                id = macroId + "_" + inc;
+                mb.setParameter(ID, id);
+            }
+            Map<String, String> lbParam = new HashMap<>();
+            lbParam.put(BLOCK_PARAM_ROLE, "presentation");
+            if (isActive) {
+                lbParam.put(BLOCK_PARAM_CLASS, "active");
+            }
+            /*
+            For now, I use the raw block for these reason:
+            - the LinkBlock generate a span element which break bootstrap tabs
+            - the word block can't contain a special char - but well this way it probably not better.
+             */
+            RawBlock linkBlock = new RawBlock(
+                String.format("<a href=\"#%s\" aria-controls=\"%s\" role=\"tab\" data-toggle=\"tab\">%s</a>",
+                    id, id, XMLUtils.escape(mb.getParameter(TAB_MACRO_PARAM_LABEL))),
+                Syntax.HTML_5_0);
```

Comment:
It's probably not the best way to do this. Do you have any idea how we can implement this in a cleaner way ?

Reply:
Following [the discussion that we had on matrix](https://matrix.to/#/!ikPtGZaGWtyblizzlR:matrix.xwiki.com/$17302157451120tahEz:matrix.xwiki.com?via=matrix.org&via=matrix.xwiki.com&via=mtrx.nz) it seem not really possible work around the span element. So the 2 solution that we have, are:
- keep the RawBlock element but it's not ideal as we lost some xwiki features.
- fix the issue, when there are a span around the link.

After a test, the issue is releated to the CSS part. The JS part still work if we use link block.

After some investiguation, it's probably possible to fix the theme to make it working with span but it might be difficult to support all flaming theme this way as in some theme there are customization for tab element.

On other side, the tab element without span is used at many place in XWiki, so to me the argument that we lost xwiki features if we don't wrap link in a span is not really important (if it was really an issue we should fix all tab in xwiki).
So my conclusion would be to keep it as it is with the rawBlock, which is not ideal but is the most simple solution in this context.

Reply:
Lets keep it as it is then. Thank u for the investigation